Support NumPy 1.13

notoraptor opened the pull request:
fix #6217 .

NumPy versions prior to 1.13 should print a `FutureWarning` when a comparaison `array == None` is done:
```
(test) C:\Users\notoraptor\mila\dev\git\theano>python -c "import numpy; print(numpy.__version__); a = numpy.arange(10); print(a == None)"
1.12.1
-c:1: FutureWarning: comparison to `None` will result in an elementwise object comparison in the future.
False
```

So, as a first check, I looked for `FutureWarning` into a previous successful CI testing, in both Jenkins and Travis:
- http://earlgrey.iro.umontreal.ca:8080/job/Theano_PR/616/console
- https://travis-ci.org/Theano/Theano/builds/263080988

As results:
- There are no `FutureWarning` messages in Jenkins console output.
- There are 4 `FutureWarning` messages in Travis tests, but all come frome SciPy (not Theano), via the same test `theano.sparse.tests.test_basic:Test_getitem.test_GetItem2D`. These are related Travis log files:
  - https://travis-ci.org/Theano/Theano/jobs/263080993
  - https://travis-ci.org/Theano/Theano/jobs/263080994
  - https://travis-ci.org/Theano/Theano/jobs/263081005
  - https://travis-ci.org/Theano/Theano/jobs/263081006

And I don't find any `== None`, `!= None`, `None ==` or `None !=` in Theano code. So, it seems that comparison `array == None` is not currently used in Theano.

What I suggest is to run Travis with NumPy 1.13, and then check if some errors occur. This PR contains a temporar commit that should install NumPy 1.13 for Travis tests. Let's see if all tests pass.

What do you think, @nouiz ?

nouiz commented:
If it is version prior to 1.13 that print warning, we should install 1.12. The old one we have probably don't have those warning.

notoraptor commented:
@nouiz I forget to mention that the `FutureWarning` from SciPy is the one related to that comparison (probably because SciPy use NumPy ? ):
```
test_GetItem2D (theano.sparse.tests.test_basic.Test_getitem) ... /home/travis/miniconda2/envs/pyenv/lib/python2.7/site-packages/scipy/sparse/csr.py:262: FutureWarning: comparison to `None` will result in an elementwise object comparison in the future.
  col.step in (1, None) and
SKIP: Slicing with step is supported.
```

So it seems the old NumPy version we use (`1.9`) already prints it.

But you are right, we should better test with NumPy `1.12`.

nouiz commented:
scipy is build on top of numpy.

notoraptor commented:
There are packages issues with NumPy/SciPy for python 3.4 with conda. It seems that conda repositories currently don't have a scipy package compatible with either numpy 1.12 or 1.13.

E.g. for 1.12:
```
(boccosetPython3) [SLURM] boccoset@leto24:~$ python -V
Python 3.4.5 :: Continuum Analytics, Inc.
(boccosetPython3) [SLURM] boccoset@leto24:~$ conda install numpy=1.12
Fetching package metadata .........
Solving package specifications: .

Package plan for installation in environment /u/boccoset/.conda/envs/boccosetPython3:

The following NEW packages will be INSTALLED:

    mkl:   2017.0.3-0
    numpy: 1.12.1-py34_0

Proceed ([y]/n)?

numpy-1.12.1-p 100% |###################################################################################################################################################################| Time: 0:00:00  41.84 MB/s
(boccosetPython3) [SLURM] boccoset@leto24:~$ conda install scipy
Fetching package metadata .........
Solving package specifications: .

Package plan for installation in environment /u/boccoset/.conda/envs/boccosetPython3:

The following NEW packages will be INSTALLED:

    libgfortran: 3.0.0-1
    scipy:       0.18.1-np111py34_1

The following packages will be DOWNGRADED due to dependency conflicts:

    numpy:       1.12.1-py34_0      --> 1.11.3-py34_0

Proceed ([y]/n)? n

Exiting
```

If you install numpy 1.12 and then you try to install scipy, conda will want to downgrade numpy back to 1.11.

I don't know if it's a temporar issue with conda, but this problem doesn't occur with python `3.5` (which hasscipy 0.19, compatible with numpy 1.12 and numpy 1.13). Maybe should we upgrade Travis Python 3 version to `3.5` (currently `3.4`) ?

@nouiz @abergeron

nouiz commented:
It was decided that travis would test the oldest version of software we
support, but that jenkins would test more recent.

for your test, you can update to 3.5, but this won't be merged in the
master.

On Mon, Aug 14, 2017 at 8:41 AM notoraptor <notifications@github.com> wrote:

> There are packages issues with NumPy/SciPy for python 3.4 with conda. It
> seems that conda repositories currently don't have a scipy package
> compatible with either numpy 1.12 or 1.13.
>
> E.g. for 1.12:
>
> (boccosetPython3) [SLURM] boccoset@leto24:~$ python -V
> Python 3.4.5 :: Continuum Analytics, Inc.
> (boccosetPython3) [SLURM] boccoset@leto24:~$ conda install numpy=1.12
> Fetching package metadata .........
> Solving package specifications: .
>
> Package plan for installation in environment /u/boccoset/.conda/envs/boccosetPython3:
>
> The following NEW packages will be INSTALLED:
>
>     mkl:   2017.0.3-0
>     numpy: 1.12.1-py34_0
>
> Proceed ([y]/n)?
>
> numpy-1.12.1-p 100% |###################################################################################################################################################################| Time: 0:00:00  41.84 MB/s
> (boccosetPython3) [SLURM] boccoset@leto24:~$ conda install scipy
> Fetching package metadata .........
> Solving package specifications: .
>
> Package plan for installation in environment /u/boccoset/.conda/envs/boccosetPython3:
>
> The following NEW packages will be INSTALLED:
>
>     libgfortran: 3.0.0-1
>     scipy:       0.18.1-np111py34_1
>
> The following packages will be DOWNGRADED due to dependency conflicts:
>
>     numpy:       1.12.1-py34_0      --> 1.11.3-py34_0
>
> Proceed ([y]/n)? n
>
> Exiting
>
> If you install numpy 1.12 and then you try to install scipy, conda will
> want to downgrade numpy back to 1.11.
>
> I don't know if it's a temporar issue with conda, but this problem doesn't
> occur with python 3.5 (which hasscipy 0.19, compatible with numpy 1.12
> and numpy 1.13). Maybe should we upgrade Travis Python 3 version to 3.5
> (currently 3.4) ?
>
> @nouiz <https://github.com/nouiz> @abergeron
> <https://github.com/abergeron>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Theano/Theano/pull/6301#issuecomment-322179580>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AALC-wcUiizAGmCgiOLvXV_ph4ijeBz7ks5sYEB2gaJpZM4O0yCo>
> .
>

notoraptor commented:
All Travis tests pass with NumPy 1.13 on Python 2.7 and 3.5 : https://travis-ci.org/notoraptor/Theano/builds/264731106 . So, it seems NumPy 1.13 works well with CPU backend.

Now let's test with NumPy 1.12 (to look for FutureWarning messages).

notoraptor commented:
All travis tests pass with numpy 1.12 on Python 2.7 and 3.5: https://travis-ci.org/notoraptor/Theano/builds/264828332

Still only 4 FutureWarning in 4 tests, with always the same warning coming from SciPy:
```
test_GetItem2D (theano.sparse.tests.test_basic.Test_getitem) ... /home/travis/miniconda2/envs/pyenv/lib/python2.7/site-packages/scipy/sparse/csr.py:301: FutureWarning: comparison to `None` will result in an elementwise object comparison in the future.
  col.step in (1, None) and
SKIP: Slicing with step is supported.
```

There are no other FutureWarning messages. So it seems comparison `array == None` is not usd in CPU backend.

Now we need to test on GPU backend.

notoraptor commented:
Run gpuarray tests with NumPy 1.12, no `FutureWarning` messages.

Script used:
```bash
#!/usr/bin/env bash

# set -e

source activate boccosetP2N112

export CUDNN=7
export NUMPY=112

export TIMESTAMP=`date +%Y-%m-%d-%Hh-%Mmin-%Ssec`

echo " "
echo "Timestamp is ${TIMESTAMP}"
echo " "
python -c "from __future__ import print_function; import numpy; print('NumPy version:', numpy.__version__)"
echo " "

cd ~/mila/dev/git/theano
git checkout master
git pull

echo CLEANING THEANO
theano-cache purge

echo "TEST gpuarray with ${CUDNN}"
echo " " >> tests-gpuarray-numpy-${NUMPY}-v${CUDNN}-${TIMESTAMP}.log

nosetests -vs theano/gpuarray/tests/ >> tests-gpuarray-numpy-${NUMPY}-v${CUDNN}-${TIMESTAMP}.log 2>&1
echo " " >> tests-gpuarray-numpy-${NUMPY}-v${CUDNN}-${TIMESTAMP}.log
echo "++++++++++++++++++++++++++++++++++++++++++++++++++" >> tests-gpuarray-numpy-${NUMPY}-v${CUDNN}-${TIMESTAMP}.log
echo " " >> tests-gpuarray-numpy-${NUMPY}-v${CUDNN}-${TIMESTAMP}.log

echo END TEST

theano-cache list > theano-cache-list-gpuarray-v${CUDNN}-${TIMESTAMP}.log
echo THEANO CACHE SAVED

source deactivate

# sbatch --gres=gpu:titanx --qos=high --mail-type=ALL --mail-user=stevenbocco@gmail.com run-numpy-112-gpuarray.sh
```

Script output:
```
Timestamp is 2017-08-21-10h-14min-11sec

NumPy version: 1.12.1

Already on 'master'
Your branch is up-to-date with 'central/master'.
Already up-to-date.
CLEANING THEANO
TEST gpuarray with 7
END TEST
THEANO CACHE SAVED
```

Output: 
[tests-gpuarray-numpy-112-v7-2017-08-21-10h-14min-11sec.log.txt](https://github.com/Theano/Theano/files/1239747/tests-gpuarray-numpy-112-v7-2017-08-21-10h-14min-11sec.log.txt)

FFT tests was skipped because I forgot to install scikit-cuda, but even with this module installed, no `FutureWarning` messages:
```
(boccosetP2N112) [SLURM] boccoset@eos18:~/mila/dev/git/theano$ nosetests -xvs theano/gpuarray/tests/test_fft.py
Using cuDNN version 7001 on context None
Mapped name None to device cuda: GeForce GTX TITAN X (0000:02:00.0)
test_1Dfft (theano.gpuarray.tests.test_fft.TestFFT) ... ok
test_grad (theano.gpuarray.tests.test_fft.TestFFT) ... ok
test_irfft (theano.gpuarray.tests.test_fft.TestFFT) ... ok
test_norm (theano.gpuarray.tests.test_fft.TestFFT) ... ok
test_odd (theano.gpuarray.tests.test_fft.TestFFT) ... ok
test_params (theano.gpuarray.tests.test_fft.TestFFT) ... ok
test_rfft (theano.gpuarray.tests.test_fft.TestFFT) ... ok
test_type (theano.gpuarray.tests.test_fft.TestFFT) ... ok

----------------------------------------------------------------------
Ran 8 tests in 141.196s

OK
```

Instead of `FutureWarning` messages, I have many occurences of the following error (not able to reproduce it yet):
```
/u/boccoset/mila/dev/git/theano/theano/tensor/var.py:526: DeprecationWarning: elementwise == comparison failed; this will raise an error in the future.
```

It does not seem to be related to comparison `array == None` (as NumPy would print a FutureWarning in such case), but I don't yet know what this warning is raised.

Now I will run gpuarray tests with NumPy **`1.13`** to check if everything works.

@nouiz @abergeron

notoraptor commented:
I made a commit to fix most of (if not all) deprecation warnings like `DeprecationWarning: elementwise == comparison failed; this will raise an error in the future.`. Jenkins used to print such warnings (e.g. previous tests for this PR: http://earlgrey.iro.umontreal.ca:8080/job/Theano_PR/652/console ), so let's see if it's printed again.

I will also run tests with special command
```
python -Wd `which nosetests` -vs ...
```
to detect all such warnings.

@abergeron @nouiz

notoraptor commented:
All tests passed with NumPy 1.13 on new backend, so everything seems to work in Theano with NumPy 1.13.

Script output:
```
Timestamp is 2017-08-21-14h-33min-22sec

NumPy version: 1.13.1

Already on 'master'
Your branch is up-to-date with 'central/master'.
Already up-to-date.
CLEANING THEANO
TEST gpuarray with 7
END TEST
THEANO CACHE SAVED
```

Output:
[tests-gpuarray-numpy-113-v7-2017-08-21-14h-33min-22sec.log.txt](https://github.com/Theano/Theano/files/1240076/tests-gpuarray-numpy-113-v7-2017-08-21-14h-33min-22sec.log.txt)


Not yet tested the last commit about deprecation warning, I will test it too.

notoraptor commented:
**NB**: As this PR contains some unwanted changes (about TRAVIS testings), it **needs** a cleanup before any merging.

notoraptor commented:
Last commit tested with NumPy 1.13 on gpuarray backend, no more warnings related to numpy comparisons.

Script output:
```
Timestamp is 2017-08-22-08h-36min-52sec

NumPy version: 1.13.1

Switched to branch 'master'
Your branch is up-to-date with 'central/master'.
error: branch 'fix-numpy-array-equals-none' not found.
Fetching origin
Fetching central
From git://github.com/Theano/Theano
   c470bd386..b998dc61d  master     -> central/master
Fetching joaovictortr
From https://github.com/joaovictortr/Theano
 + 6dc9fc7a4...266a84edb spatialtf_cpu -> joaovictortr/spatialtf_cpu  (forced update)
Switched to a new branch 'fix-numpy-array-equals-none'
Branch fix-numpy-array-equals-none set up to track remote branch fix-numpy-array-equals-none from origin.
CLEANING THEANO
TEST gpuarray with cuDNN v7
END TEST
THEANO CACHE SAVED
```

Output: 
[tests-gpuarray-numpy-113-cudnn-v7-2017-08-22-08h-36min-52sec.log.txt](https://github.com/Theano/Theano/files/1242617/tests-gpuarray-numpy-113-cudnn-v7-2017-08-22-08h-36min-52sec.log.txt)

No future warnings. There are deprecation warnings, but seems unrelated to NumPy:
```
notoraptor@DESKTOP-HVUEVJI MINGW64 ~/mila/dev/boccoset
$ grep future -i tests-gpuarray-numpy-113-cudnn-v7-2017-08-22-08h-36min-52sec.log
```
```
notoraptor@DESKTOP-HVUEVJI MINGW64 ~/mila/dev/boccoset
$ grep warn -i tests-gpuarray-numpy-113-cudnn-v7-2017-08-22-08h-36min-52sec.log
/u/boccoset/mila/dev/git/theano/theano/gpuarray/tests/test_dnn.py:656: DeprecationWarning: testcase_func_name= is deprecated; use name_func=
test_stack_new_interface (theano.gpuarray.tests.test_basic_ops.G_Join_and_Split) ... /u/boccoset/.conda/envs/boccosetP2N113/lib/python2.7/unittest/case.py:329: DeprecationWarning: stack(*tensors) interface is deprecated, use stack(tensors, axis=0) instead.
test_softmax_grad (theano.gpuarray.tests.test_dnn.test_SoftMax) ... /u/boccoset/mila/dev/git/theano/theano/gpuarray/tests/test_dnn.py:1351: UserWarning: DEPRECATION: If x is a vector, Softmax will not automatically pad x anymore in next releases. If you need it, please do it manually. The vector case is gonna be supported soon and the output will be a vector.
/u/boccoset/mila/dev/git/theano/theano/gpuarray/tests/test_dnn.py:1377: UserWarning: DEPRECATION: If x is a vector, Softmax will not automatically pad x anymore in next releases. If you need it, please do it manually. The vector case is gonna be supported soon and the output will be a vector.
test_infer_shape (theano.gpuarray.tests.test_elemwise.test_GpuCAReduceCuda) ... WARNING (theano.tests.unittest_tools): While testing shape inference for <class 'theano.gpuarray.elemwise.GpuCAReduceCuda'>, we received an input with a shape that has some repeated values: (128, 1, 3, 3), like a square matrix. This makes it impossible to check if the values for these dimensions have been correctly used, or if they have been mixed up.
WARNING (theano.tests.unittest_tools): While testing shape inference for <class 'theano.gpuarray.elemwise.GpuCAReduceCuda'>, we received an input with a shape that has some repeated values: (128, 1, 3, 3), like a square matrix. This makes it impossible to check if the values for these dimensions have been correctly used, or if they have been mixed up.
Tests that multinomial_wo_replacement fails when asked to sample more ... /u/boccoset/mila/dev/git/theano/theano/sandbox/rng_mrg.py:1020: UserWarning: MRG_RandomStreams.multinomial_wo_replacement() is deprecated and will be removed in the next release of Theano. Please use MRG_RandomStreams.choice() instead.
  warnings.warn('MRG_RandomStreams.multinomial_wo_replacement() is '
test_elemwise_fusion (theano.gpuarray.tests.test_others.test_fusion) ... /u/boccoset/mila/dev/git/theano/theano/tensor/basic.py:2117: UserWarning: theano.tensor.round() changed its default from `half_away_from_zero` to `half_to_even` to have the same default as NumPy. Use the Theano flag `warn.round=False` to disable this warning.
theano.gpuarray.tests.test_pickle.test_unpickle_gpuarray_as_numpy_ndarray_flag2 ... /u/boccoset/mila/dev/git/theano/theano/gpuarray/type.py:861: UserWarning: config.experimental.unpickle_gpu_on_cpu is set to True. Unpickling GpuArray as numpy.ndarray
```
I have also checked last TRAVIS tests (currently run with numpy 1.12), only expected 4 `FutureWarning`, no deprecation warnings about failed elemwise comparison.

So, it seems NumPy 1.13 is ready to recommend with Theano. I will clean this PR so that we can merge the fix about deprecation warning.

@nouiz @abergeron

notoraptor commented:
Useless changes removed. Just kept fixes in `ftheano/tensor/vars.py`.

@nouiz @abergeron

notoraptor commented:
And rebased to remove dummy commits.

nouiz commented:
travis fail

notoraptor commented:
Error:
```
Warning, treated as error:

WARNING: **********************************************************************

File "tutorial/gradients.txt", line 137, in default

Failed example:

    f = theano.function([x], J, updates=updates)

Exception raised:

    Traceback (most recent call last):

      File "/home/travis/miniconda2/envs/pyenv/lib/python2.7/doctest.py", line 1315, in __run

        compileflags, 1) in test.globs

      File "<doctest default[5]>", line 1, in <module>

        f = theano.function([x], J, updates=updates)

      File "/home/travis/build/Theano/Theano/theano/compile/function.py", line 317, in function

        output_keys=output_keys)

      File "/home/travis/build/Theano/Theano/theano/compile/pfunc.py", line 486, in pfunc

        output_keys=output_keys)

      File "/home/travis/build/Theano/Theano/theano/compile/function_module.py", line 1838, in orig_function

        fn = m.create(defaults)

      File "/home/travis/build/Theano/Theano/theano/compile/function_module.py", line 1712, in create

        input_storage=input_storage_lists, storage_map=storage_map)

      File "/home/travis/build/Theano/Theano/theano/gof/link.py", line 699, in make_thunk

        storage_map=storage_map)[:3]

      File "/home/travis/build/Theano/Theano/theano/gof/vm.py", line 1084, in make_all

        impl=impl))

      File "/home/travis/build/Theano/Theano/theano/scan_module/scan_op.py", line 925, in make_thunk

        from . import scan_perform_ext

      File "/home/travis/build/Theano/Theano/theano/scan_module/scan_perform_ext.py", line 149, in <module>

        from scan_perform.scan_perform import *

      File "__init__.pxd", line 155, in init theano.scan_module.scan_perform (/home/travis/.theano/compiledir_Linux-4.11--generic-x86_64-with-debian-jessie-sid-x86_64-2.7.13-64/scan_perform/mod.cpp:10916)

    ValueError: ('The following error happened while compiling the node', for{cpu,scan_fn}(Shape_i{0}.0, Subtensor{int64:int64:int8}.0, Shape_i{0}.0, Elemwise{sqr,no_inplace}.0, x, Alloc.0), '\n', 'numpy.dtype has the wrong size, try recompiling. Expected 88, got 96')
```

Is it because it tries to go back from NumPy 1.12 to NumPy 1.9 (I have not modified any C code) ? Should I try to clear TRAVIS cache ?

notoraptor commented:
I deleted cache for this PR and restarted Travis.
